File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        ans = 0
        n = len(nums)

        for i in range(n):
            left, right = i + 1, n - 1

            while left <= right:
                mid = (left + right) // 2
                if nums[i] + nums[mid] < lower:
                    left = mid + 1
                else:
                    right = mid - 1

            j = left

            left, right = i + 1, n - 1
            while left <= right:
                mid = (left + right) // 2
                if nums[i] + nums[mid] <= upper:
                    left = mid + 1
                else:
                    right = mid - 1

            k = left
            ans += k - j

        return ans

File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.cpp
class Solution {
public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(), nums.end());
        return countLessEqual(nums, upper) - countLessEqual(nums, lower - 1);
    }

private:
    long long countLessEqual(vector<int>& nums, int val) {
        long long count = 0;
        int left = 0, right = nums.size() - 1;
        
        while (left < right) {
            int sum = nums[left] + nums[right];
            if (sum <= val) {
                count += right - left;
                left++;
            } else {
                right--;
            }
        }
        return count;
    }
};